File: 78-subsets/subsets.java
class Solution {
    public void recursiveSubsets(int[] nums, List<Integer> path, List<List<Integer>> result, int i) 
    {
        if (i == nums.length)
        {
            result.add(new ArrayList<>(path));
            return;
        }
        // include nums[i]
        path.add(nums[i]);
        recursiveSubsets(nums, path, result, i + 1);
        path.remove(path.size() - 1);

        // exclude nums[i]
        recursiveSubsets(nums, path, result, i + 1);
    }

    public List<List<Integer>> subsets(int[] nums) 
    {
        List<Integer> path = new ArrayList<>();
        List<List<Integer>> result = new ArrayList<>();
        recursiveSubsets(nums, path, result, 0);
        return result;
    }
}

File: 78-subsets/subsets.py
class Solution:
    def recursive_subsets(self, nums: list[int], path: list[int], result: list[list[int]], i: int) -> None:
        if i == len(nums):
            result.append(path.copy())
            return

        path.append(nums[i])
        self.recursive_subsets(nums, path, result, i + 1)
        path.pop()

   
        self.recursive_subsets(nums, path, result, i + 1)

    def subsets(self, nums: list[int]) -> list[list[int]]:
        path = []
        result = []
        self.recursive_subsets(nums, path, result, 0)
        return result